contracts/UpgradeManager.sol

```diff
@@ -0,0 +1,405 @@
+    );
+
+    // This uses staticcall because there is no way to just get a false return
+    // if the result is negative, the transaction reverts, so we detect that and
+    // give a better error message
+    (bool success, bytes memory returnData) = _proxyAdminAddress.staticcall(
+      abi.encodeWithSignature("getProxyAdmin(address)", _proxyAddress)
+    );
+
+    require(success, "Call to determine proxy admin ownership of proxy failed");
+
+    address returnedAddress = abi.decode(returnData, (address));
+
+    require(
+      returnedAddress == _proxyAdminAddress,
+      "Proxy admin is not admin of this proxy"
+    );
+  }
+
+  function _addUpgradeProposer(address proposerAddress) private {
+    upgradeProposers.add(proposerAddress);
+    emit ProposerAdded(proposerAddress);
+  }
+
+  function _applyChanges(address _proxyAddress) private {
+    AdoptedContract storage adoptedContract = adoptedContractsByProxyAddress[
+      _proxyAddress
+    ];
+
+    if (adoptedContract.upgradeAddress == address(0)) {
+      // solhint-disable-next-line avoid-low-level-calls
+      (bool success, ) = _proxyAddress.call(adoptedContract.encodedCall);
+      require(success, "call failed");
+    } else {
+      IProxyAdmin proxyAdmin = IProxyAdmin(adoptedContract.proxyAdmin);
+
+      if (adoptedContract.encodedCall.length == 0) {
+        proxyAdmin.upgrade(_proxyAddress, adoptedContract.upgradeAddress);
+      } else {
+        proxyAdmin.upgradeAndCall(
+          _proxyAddress,
+          adoptedContract.upgradeAddress,
+          adoptedContract.encodedCall
+        );
+      }
+    }
+  }
+
+  function _propose(
+    string calldata _contractId,
+    address _implementationAddress,
+    bytes memory encodedCall
+  ) private {
+    address proxyAddress = adoptedContractAddresses[_contractId];
+
+    require(proxyAddress != address(0), "Unknown contract id");
+
+    require(
+      !proxiesWithPendingChanges.contains(proxyAddress),
+      "Upgrade already proposed, withdraw first"
+    );
+
+    AdoptedContract storage adoptedContract = adoptedContractsByProxyAddress[
+      proxyAddress
+    ];
+    address currentImplementationAddress = IProxyAdmin(
+      adoptedContract.proxyAdmin
+    ).getProxyImplementation(proxyAddress);
+
+    require(
+      currentImplementationAddress != _implementationAddress,
+      "Implementation address unchanged"
+    );
+
+    if (_implementationAddress != address(0)) {
+      // Note: isContract() is not guaranteed to return an accurate value, never use it to provide an assurance of security, this
+      // is just a last line of defence against footgun
+      require(
+        AddressUpgradeable.isContract(_implementationAddress),
+        "Implementation address is not a contract"
+      );
+    }
+
+    proxiesWithPendingChanges.add(proxyAddress);
+
+    adoptedContract.upgradeAddress = _implementationAddress;
+    adoptedContract.encodedCall = encodedCall;
+
+    nonce++;
```

Comment:
I think it is and I think this means I should comment here.

Here is the point of the nonce:

Imagine there is an evil proposer, maybe they stole a private key from a legit developer.

They could propose a bad change, but hopefully the owner will review all the changes before they call `upgradeProtocol`.

However, with no nonce, the attacker can watch for that `upgradeProtocol` call, and when they see it quickly submit a proposal for evil code that gets in the same block, and therefore the code is included even though the owner never saw it before submitting their transaction.

The nonce is so that the owner can know exactly the set of changes they are proposing when they call upgradeProtocol - the nonce is checked [here](https://github.com/cardstack/card-pay-protocol/pull/234/files/cb296928f16ae601a950e2e9e5a2cf0ecd32d21c#diff-4fee80438489cc7e7ede1017507a882df004aea23b5377d563e18c1e1d26c5cbR176)

Does that make sense, or have I missed something else? If it makes sense I will add a comment explaining the nonce purpose

Reply:
I see, yes it makes sense